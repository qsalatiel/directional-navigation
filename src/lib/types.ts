export enum Directions {
  LEFT = "left",
  RIGHT = "right",
  UP = "up",
  DOWN = "down",
}

export enum KeyCodes {
  LEFT = 37,
  UP = 38,
  RIGHT = 39,
  DOWN = 40,
}

export type SectionDirections = "horizontal" | "vertical" | "grid";

export type FocusPositions = {
  isFocusInFirstColumn: boolean;
  isFocusInLastColumn: boolean;
  isFocusInFirstRow: boolean;
  isFocusInLastRow: boolean;
};

export type GetGridFocusPositionsParams = {
  index: number;
  rows: number;
  columns: number;
};
