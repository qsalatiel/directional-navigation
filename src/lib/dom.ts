import { DIRECTION_NAVIGATE_ATTR, PARENT_SECTION_ATTR } from '../constants'
import { Directions } from './types'

export function qs(selector: string): HTMLElement | null {
  return document.querySelector(selector)
}

export function all(selector: string): NodeList {
  return document.querySelectorAll(selector)
}

export function focusElement($el?: HTMLElement | null) {
  if (!$el) return console.error('unable to focus element', $el)

  if (typeof $el.focus === 'function') {
    $el.focus()
  }
}

export function getFocusedElement() {
  return document.activeElement as HTMLElement | null
}

export function getParentSectionId($el: Element) {
  return $el.getAttribute(PARENT_SECTION_ATTR)
}

export function getNextElementFromDirection(
  $el: Element,
  direction: Directions
): HTMLElement | null {
  const directionAttribute = DIRECTION_NAVIGATE_ATTR.replace('{direction}', direction)

  const nextElementSelector = $el.getAttribute(directionAttribute)

  if (nextElementSelector) {
    return qs(nextElementSelector)
  }

  return null
}
