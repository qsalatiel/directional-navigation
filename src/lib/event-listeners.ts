import { handleFocusEvent, handleKeyupEvent } from './nagivation'

export function listenKeyboardEvents() {
  window.addEventListener('keydown', handleKeyupEvent)
}

export function removeKeyboardEvents() {
  window.removeEventListener('keydown', handleKeyupEvent)
}

export function listenFocusEvent() {
  window.addEventListener('focusin', handleFocusEvent)
}

export function removeFocusEvent() {
  window.removeEventListener('focusin', handleFocusEvent)
}
