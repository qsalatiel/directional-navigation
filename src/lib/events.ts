import { DID_FOCUS_EVENT, LIB_DESTROYED_EVENT, LIB_INITIALIZED_EVENT } from '../constants'

export function dispatchFocusEvent($el: HTMLElement) {
  const event = new CustomEvent(DID_FOCUS_EVENT, {
    bubbles: true,
    detail: { element: $el },
  })

  $el.dispatchEvent(event)
}

export function dispatchInitEvent() {
  const event = new CustomEvent(LIB_INITIALIZED_EVENT, {
    bubbles: true,
  })

  window.dispatchEvent(event)
}

export function dispatchDestroyEvent() {
  const event = new CustomEvent(LIB_DESTROYED_EVENT, {
    bubbles: true,
  })

  window.dispatchEvent(event)
}
