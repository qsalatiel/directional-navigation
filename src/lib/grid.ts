import {
  FIRST_COLUMN_PRODUCT_RESULT,
  LAST_COLUMN_PRODUCT_RESULT,
} from "./constants";

import { GetGridFocusPositionsParams } from "./types";

export function isGridAttrValid(attr: string | null) {
  if (!attr) return false;

  if (+attr < 0) return false;

  return Number.isInteger(+attr);
}

export function getGridFocusPositions({
  index,
  rows,
  columns,
}: GetGridFocusPositionsParams) {
  const gridIndex = index + 1;

  const lastIndex = rows * columns;
  const lastIndexInSecondLastRow = lastIndex - columns;

  const isFocusInFirstColumn =
    gridIndex % columns === FIRST_COLUMN_PRODUCT_RESULT;

  const isFocusInLastColumn =
    gridIndex % columns === LAST_COLUMN_PRODUCT_RESULT;

  const isFocusInFirstRow = gridIndex <= rows;

  const isFocusInLastRow =
    gridIndex <= lastIndex && gridIndex > lastIndexInSecondLastRow;

  return {
    isFocusInFirstColumn,
    isFocusInLastColumn,
    isFocusInFirstRow,
    isFocusInLastRow,
  };
}
