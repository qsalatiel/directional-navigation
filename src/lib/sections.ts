import {
  LEAVE_SECTION_DIRECTION_ATTR,
  PARENT_SECTION_ATTR,
  SECTION_CURRENT_FOCUSED_CHILDREN_ATTR,
  SECTION_GRID_COLUMN_ATTR,
  SECTION_GRID_ROW_ATTR,
  SECTION_ID_ATTR,
} from '../constants'

import { all, getParentSectionId, qs } from './dom'
import { handleFocusElement } from './nagivation'
import { Directions, SectionDirections } from './types'
import { getGridFocusPositions, isGridAttrValid } from './grid'

export class ManagedDirectionalSection {
  sectionId: string
  $sectionEl: HTMLElement
  chidlren: NodeList
  currentFocusedChildrenIndex: number
  direction: SectionDirections
  gridRows?: number
  gridColumns?: number

  constructor(sectionId: string, direction: SectionDirections) {
    this.sectionId = sectionId
    this.direction = direction

    const $sectionEl = qs(`[${SECTION_ID_ATTR}='${sectionId}']`)

    if (!$sectionEl)
      throw new Error(
        `Error: Unable to instantiate a section. Tried to querySelector([${SECTION_ID_ATTR}=${sectionId}]) but not found any match element.`
      )

    this.$sectionEl = $sectionEl

    this.chidlren = ManagedDirectionalSection.getChildren(sectionId)

    const childrenIndex = $sectionEl.getAttribute(SECTION_CURRENT_FOCUSED_CHILDREN_ATTR)

    this.currentFocusedChildrenIndex = childrenIndex ? Number(childrenIndex) : 0

    if (direction === 'grid') {
      const gridRows = $sectionEl.getAttribute(SECTION_GRID_ROW_ATTR)
      const gridColumns = $sectionEl.getAttribute(SECTION_GRID_COLUMN_ATTR)

      if (
        !gridRows ||
        !gridColumns ||
        !isGridAttrValid(gridRows) ||
        !isGridAttrValid(gridColumns)
      ) {
        throw new Error(
          `Tried to instantiate a grid section with invalid columns or rows. Received: data-grid-rows="${gridRows}" and data-grid-columns="${gridColumns}". Both are required. Please add a integer number greater than 0.`
        )
      }

      this.gridRows = +gridRows
      this.gridColumns = +gridColumns
    }
  }

  static getElementSection($el: HTMLElement) {
    const parentSectionId = getParentSectionId($el)

    if (!parentSectionId) return null

    const section = getSection(parentSectionId)

    if (!section) return null

    return section
  }

  static getChildren(sectionId: string) {
    const getChildrenList = () => all(`[${PARENT_SECTION_ATTR}='${sectionId}']`)

    return getChildrenList()
  }

  setCurrentFocusedChildrenIndex(index: number) {
    this.$sectionEl.setAttribute(SECTION_CURRENT_FOCUSED_CHILDREN_ATTR, String(index))

    this.currentFocusedChildrenIndex = index
  }

  focusChildren($el: HTMLElement) {
    const nextChildrenFocus = Array.prototype.indexOf.call(this.chidlren, $el)

    if (nextChildrenFocus > 0) {
      this.setCurrentFocusedChildrenIndex(nextChildrenFocus)
    }
  }

  leaveSection(direction: Directions) {
    const leaveToSelector = this.$sectionEl.getAttribute(
      `${LEAVE_SECTION_DIRECTION_ATTR.replace('{direction}', direction)}`
    )

    if (!leaveToSelector) return

    const leaveToElement = qs(leaveToSelector)

    if (!leaveToElement) return

    handleFocusElement(leaveToElement)
  }

  handleGridFocus(keyPressed: Directions) {
    if (!this.gridColumns || !this.gridRows) return

    const index = this.currentFocusedChildrenIndex

    const { isFocusInFirstColumn, isFocusInLastColumn, isFocusInFirstRow, isFocusInLastRow } =
      getGridFocusPositions({
        index,
        rows: this.gridRows,
        columns: this.gridColumns,
      })

    switch (keyPressed) {
      case Directions.RIGHT:
        return isFocusInLastColumn
          ? this.leaveSection(Directions.RIGHT)
          : this.focusIndex(index + 1)

      case Directions.LEFT:
        return isFocusInFirstColumn
          ? this.leaveSection(Directions.LEFT)
          : this.focusIndex(index - 1)

      case Directions.UP:
        return isFocusInFirstRow
          ? this.leaveSection(Directions.UP)
          : this.focusIndex(index - this.gridColumns)

      case Directions.DOWN:
        return isFocusInLastRow
          ? this.leaveSection(Directions.DOWN)
          : this.focusIndex(index + this.gridColumns)
      default:
        return
    }
  }

  focusFromKeyPressed(keyPressed: Directions) {
    if (this.direction === 'grid') {
      this.handleGridFocus(keyPressed)
    }

    if (this.direction === 'horizontal') {
      if (keyPressed === Directions.RIGHT) {
        return this.focusNext()
      }

      if (keyPressed === Directions.LEFT) {
        return this.focusPrevious()
      }

      return this.leaveSection(keyPressed)
    }

    if (this.direction === 'vertical') {
      if (keyPressed === Directions.DOWN) {
        return this.focusNext()
      }

      if (keyPressed === Directions.UP) {
        return this.focusPrevious()
      }

      return this.leaveSection(keyPressed)
    }
  }

  focusPrevious() {
    const previousFocusIndex = this.currentFocusedChildrenIndex - 1

    if (previousFocusIndex >= 0) {
      const previousFocusElement = this.chidlren.item(previousFocusIndex) as HTMLElement | null

      if (!previousFocusElement) return

      this.setCurrentFocusedChildrenIndex(previousFocusIndex)

      return handleFocusElement(previousFocusElement)
    }

    if (this.direction === 'horizontal') {
      this.leaveSection(Directions.LEFT)
    }

    if (this.direction === 'vertical') {
      this.leaveSection(Directions.UP)
    }
  }

  focusNext() {
    const nextFocusIndex = this.currentFocusedChildrenIndex + 1

    if (nextFocusIndex <= this.chidlren.length) {
      const nextFocusElement = this.chidlren.item(nextFocusIndex) as HTMLElement | null

      if (!nextFocusElement) return

      this.setCurrentFocusedChildrenIndex(nextFocusIndex)

      return handleFocusElement(nextFocusElement)
    }

    if (this.direction === 'horizontal') {
      this.leaveSection(Directions.RIGHT)
    }

    if (this.direction === 'vertical') {
      this.leaveSection(Directions.DOWN)
    }
  }

  focusIndex(index: number) {
    const focusCandidate = this.chidlren.item(index) as HTMLElement | null

    if (focusCandidate) {
      handleFocusElement(focusCandidate)
      this.setCurrentFocusedChildrenIndex(index)
    }
  }

  focusLastFocusedElement() {
    const lastCurrentFocusedElement = this.chidlren.item(
      this.currentFocusedChildrenIndex
    ) as HTMLElement | null

    if (!lastCurrentFocusedElement) return

    handleFocusElement(lastCurrentFocusedElement)
  }
}

export function getSection(sectionId: string) {
  if (!window.__dn_sections__) return

  const section = window.__dn_sections__[sectionId] as ManagedDirectionalSection | undefined

  return section
}
