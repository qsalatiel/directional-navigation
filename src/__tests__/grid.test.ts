import { getGridFocusPositions, isGridAttrValid } from '../lib/grid'

describe('grid functions', () => {
  it('should result in a valid grid attr', () => {
    expect(isGridAttrValid('9')).toBeTruthy()
  })

  it('should get false passing null as grid attr', () => {
    expect(isGridAttrValid(null)).toBeFalsy()
  })

  it('should get false passing -1 as grid attr', () => {
    expect(isGridAttrValid('-1')).toBeFalsy()
  })

  it('should get false passing 3.14 as grid attr', () => {
    expect(isGridAttrValid('3.14')).toBeFalsy()
  })

  it('should match snapshot passing index: 0, ', () => {
    const result = getGridFocusPositions({
      index: 0,
      columns: 5,
      rows: 5,
    })

    expect(result).toMatchSnapshot()
  })

  it('should match snapshot passing index: 3, ', () => {
    const result = getGridFocusPositions({
      index: 3,
      columns: 5,
      rows: 5,
    })

    expect(result).toMatchSnapshot()
  })

  it('should match snapshot passing index: 21, ', () => {
    const result = getGridFocusPositions({
      index: 21,
      columns: 5,
      rows: 5,
    })

    expect(result).toMatchSnapshot()
  })

  it('should match snapshot passing index: 24, ', () => {
    const result = getGridFocusPositions({
      index: 24,
      columns: 5,
      rows: 5,
    })

    expect(result).toMatchSnapshot()
  })

  it('should match snapshot passing index: 13, ', () => {
    const result = getGridFocusPositions({
      index: 13,
      columns: 5,
      rows: 5,
    })

    expect(result).toMatchSnapshot()
  })
})
