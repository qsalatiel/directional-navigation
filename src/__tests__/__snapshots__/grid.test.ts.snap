// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grid functions should match snapshot passing index: 0,  1`] = `
{
  "isFocusInFirstColumn": true,
  "isFocusInFirstRow": true,
  "isFocusInLastColumn": false,
  "isFocusInLastRow": false,
}
`;

exports[`grid functions should match snapshot passing index: 3,  1`] = `
{
  "isFocusInFirstColumn": false,
  "isFocusInFirstRow": true,
  "isFocusInLastColumn": false,
  "isFocusInLastRow": false,
}
`;

exports[`grid functions should match snapshot passing index: 13,  1`] = `
{
  "isFocusInFirstColumn": false,
  "isFocusInFirstRow": false,
  "isFocusInLastColumn": false,
  "isFocusInLastRow": false,
}
`;

exports[`grid functions should match snapshot passing index: 21,  1`] = `
{
  "isFocusInFirstColumn": false,
  "isFocusInFirstRow": false,
  "isFocusInLastColumn": false,
  "isFocusInLastRow": true,
}
`;

exports[`grid functions should match snapshot passing index: 24,  1`] = `
{
  "isFocusInFirstColumn": false,
  "isFocusInFirstRow": false,
  "isFocusInLastColumn": true,
  "isFocusInLastRow": true,
}
`;
